beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/DepositProvider.java

```diff
@@ -121,6 +132,19 @@ public void onNewFinalizedCheckpoint(
           && depositMerkleTree.getDepositCount()
               >= finalizedState.getEth1Data().getDepositCount().longValue()) {
         depositMerkleTree.finalize(finalizedState.getEth1Data());
+        eth1DataCache
+            .getFirstDataEx()
+            .ifPresent(
+                eth1DataEx -> {
+                  // TODO: log
+                  final StoreTransaction transaction = recentChainData.startStoreTransaction();
+                  transaction.setEth1Cache(
+                      Eth1Cache.SSZ_SCHEMA.create(
+                          depositMerkleTree.getSnapshot().asSerializable(),
+                          eth1DataEx.getBlockNumber()),
+                      fromOptimisticBlock);
+                  transaction.commit().join();
```

Comment:
I'd be very keen to avoid the `.join()` here. I'm not entirely sure why the `eth1DataCache` is involved here at all though - it's quite separate to the deposit merkle tree and has a different life cycle for what it holds onto (based on timestamps).  There's no actual connection between the deposit we finalize which is tied to the state of the beacon chain and what we have in the eth1 data cache which is primarily tied to the time of eth1 blocks.

Reply:
I also don't think we need to run this through the `StoreTransaction` - we can just send the merkle tree snapshot to the database directly and not wait for it to complete.  By the time we get a new finalized checkpoint here, the store updates have been persisted so we're safe to prune the merkle tree and send the new snapshot to disk.  If that new snapshot fails to write for some reason, that's not a problem either - we'll just use an older snapshot when we next start up and process more deposits than we need to.